File: day_16/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <memory>
#include <cassert>
#include <regex>
#include <fstream>
#include <memory>
#include <stdint.h>

using namespace std;


struct Valve {
    string label{};
    int flow_rate{};
    vector<string> tunnels{};
};

template<class T>
class SolutionBlock {
public:
    SolutionBlock(size_t num_valves, size_t num_possible_open_valves)
        : num_valves(num_valves)
        , num_possible_open_valves(num_possible_open_valves)
    {
        size_t blocksize = 26 * num_valves * num_valves * num_possible_open_valves;
        solutions = new T[blocksize];
        assert(solutions != nullptr);
        memset(solutions, 0xFF, blocksize * sizeof(T));
    }

    ~SolutionBlock() {
        delete[] solutions;
    }

    class SolutionPtr {
    public:
        SolutionPtr(T* solution)
            : solution(solution)
        {}

        T& operator*() {
            return *solution;
        }

        void maybe_update(T value) {
            assert(value != 0xffff);
            if(value > *solution) {
                *solution = value;
            }
        }
        T* solution{nullptr};
    };

    SolutionPtr index(size_t minute, size_t uvalve_idx, size_t evalve_idx, size_t open_valves_idx) {
        T* ptr = solutions + (minute * num_valves * num_valves * num_possible_open_valves)
        + (uvalve_idx * num_valves * num_possible_open_valves)
        + (evalve_idx * num_possible_open_valves)
        + open_valves_idx;
        return SolutionPtr(ptr);
    }

private:
    size_t num_valves{};
    size_t num_possible_open_valves{};
    T* solutions{nullptr};
};

typedef shared_ptr<Valve> vptr;

vector<string> split_tunnels(const string& tunnels) {
    vector<string> out;
    unsigned long start = 0;
    unsigned long end = 0;
    do {
        end = tunnels.find(", ", start);
        out.push_back(tunnels.substr(start, end - start));
        start = end + 2;
    } while(end != string::npos);
    return out;
}

vector<vptr> read_valves(const char* infile) {
    vector<vptr> valves;
    std::regex valve_regex(R"~~(Valve (\S+) has flow rate=(\d+); tunnels? leads? to valves? (.*))~~");
    ifstream inf;
    inf.open(infile);
    assert(inf.is_open());
    string line;

    smatch matches;
    while(getline(inf, line)) {
        if(!regex_match(line, matches, valve_regex)) {
            cout << "Bad valve: \"" << line << "\"" << endl;
            continue;
        }

        auto valve = make_shared<Valve>();
        valve->label = matches[1];
        valve->flow_rate = stoi(matches[2]);
        valve->tunnels = move(split_tunnels(matches[3]));

        valves.push_back(valve);
    }

    return valves;
}

void print_valves(const vector<vptr>& valves) {
    for(auto& valve : valves) {
        cout << "Valve " << valve->label << " has flow rate=" << valve->flow_rate << "; tunnels lead to valves ";
        for(unsigned long i = 0; i < valve->tunnels.size(); ++i) {
            cout << valve->tunnels[i];
            if(i < valve->tunnels.size()-1) {
                cout << ", ";
            }
        }
        cout << endl;
    }
}

vector<vector<vptr>> powerset(const vector<vptr>& valves) {
    vector<vector<vptr>> retval;
    if(valves.size() == 0) {
        retval.push_back(vector<vptr>());
        return retval;
    }

    vector<vptr> smaller_valves = vector<vptr>(valves.begin(), valves.end()-1);
    vector<vector<vptr>> smaller_powerset = powerset(smaller_valves);

    vptr last_item = valves.back();
    for(auto& set: smaller_powerset) {
        vector<vptr> setcpy = set;
        setcpy.push_back(last_item);
        retval.push_back(set);
        retval.push_back(setcpy);
    }

    return retval;
}

inline uint16_t valve_to_bit(map<string, size_t>& valve_map, const vptr& valve) {
    size_t pos = valve_map[valve->label];
    uint16_t out = 1 << pos;
    assert(out != 0);
    return out;
}

int main(int argc, char* argv[]) {
    assert(argc >= 2);
    cout << "Reading valves" << endl;
    auto valves = read_valves(argv[1]);
    map<string, size_t> valve_map;
    for(size_t i = 0; i < valves.size(); ++i) {
        valve_map.insert(pair<string,size_t>(valves[i]->label, i));
    }

    cout << "Identifying significant valves" << endl;
    vector<vptr> significant_valves;
    for(auto& valve : valves) {
        if(valve->flow_rate > 0) {
            significant_valves.push_back(valve);
        }
    }

    cout << "Collating significant valves" << endl;
    map<string, size_t> significant_valve_map;
    for(size_t i = 0; i < significant_valves.size(); ++i) {
        significant_valve_map.insert(pair<string, size_t>(significant_valves[i]->label, i));
    }

    cout << "Calculating possible open valve combinations" << endl;
    auto possible_open_valves = powerset(significant_valves);
    vector<uint16_t> possible_open_valves_ints;
    for(auto& valve_set : possible_open_valves) {
        uint16_t valve_set_int = 0;
        for(auto& valve : valve_set) {
            uint16_t valve_bit = valve_to_bit(significant_valve_map, valve);
            assert((valve_set_int & valve_bit) == 0);
            valve_set_int |= valve_bit;
        }
        possible_open_valves_ints.push_back(valve_set_int);
    }
    cout << "Collating possible open valve combinations" << endl;
    map<uint16_t, size_t> possible_open_valves_idx_map;
    for(size_t i = 0; i < possible_open_valves_ints.size(); ++i) {
        possible_open_valves_idx_map.insert(pair<uint16_t, size_t>(possible_open_valves_ints[i], i));
    }

#if 0
    cout << "All valves:" << endl;
    print_valves(valves);
    cout << endl;

    cout << "Significant valves:" << endl;
    print_valves(significant_valves);
    cout << endl;

    cout << "Possible open valve combos:" << endl;
    for(auto& valve_set : possible_open_valves) {
        cout << "[";
        for(unsigned long i = 0; i < valve_set.size(); ++i) {
            cout << valve_set[i]->label;
            if(i < valve_set.size() - 1) {
                cout << ",";
            }
        }
        cout << "]" << endl;
    }
#endif

    // solutions[sol_idx][uvalve][evalve][open_valves] -> score
    SolutionBlock<uint16_t> solutions(valves.size(), possible_open_valves.size());

    for(size_t minute = 0; minute < 26; ++minute) {
        for(size_t uvalve_idx = 0; uvalve_idx < valves.size(); ++uvalve_idx) {
            cout << "Progress: " << ((minute * valves.size()) + uvalve_idx) << "/" << (26 * valves.size()) << endl;
            auto& uvalve = valves[uvalve_idx];
            uint16_t uvalve_bit = (uvalve->flow_rate > 0) ? valve_to_bit(significant_valve_map, uvalve) : 0;
            for(size_t evalve_idx = 0; evalve_idx < valves.size(); ++evalve_idx) {
                auto& evalve = valves[evalve_idx];
                uint16_t evalve_bit = (evalve->flow_rate > 0) ? valve_to_bit(significant_valve_map, evalve) : 0;
                for(size_t open_valves_idx = 0; open_valves_idx < possible_open_valves_ints.size(); ++open_valves_idx) {
                    uint16_t open_valves = possible_open_valves_ints[open_valves_idx];
                    auto solution = solutions.index(minute, uvalve_idx, evalve_idx, open_valves_idx);
                    assert(*solution == 0xffff);
                    *solution = 0;
                    if(minute > 0) {
                        // 1. you move, elephant moves
                        for(auto& ut : uvalve->tunnels) {
                            for(auto& et : evalve->tunnels) {
                                solution.maybe_update(*solutions.index(minute-1, valve_map[ut], valve_map[et], open_valves_idx));
                            }
                        }

                        // 2. you move, elephant opens
                        if(evalve->flow_rate > 0 && (evalve_bit & open_valves) == 0) {
                            for(auto& ut : uvalve->tunnels) {
                                uint16_t new_open_valves = open_valves | evalve_bit;
                                size_t new_open_valves_idx = possible_open_valves_idx_map[new_open_valves];
                                solution.maybe_update(*solutions.index(minute-1, valve_map[ut], evalve_idx, new_open_valves_idx) + (evalve->flow_rate * minute));
                            }
                        }

                        // 3. you open, elephant moves
                        if(uvalve->flow_rate > 0 && (uvalve_bit & open_valves) == 0) {
                            for(auto& et : evalve->tunnels) {
                                uint16_t new_open_valves = open_valves | uvalve_bit;
                                size_t new_open_valves_idx = possible_open_valves_idx_map[new_open_valves];
                                solution.maybe_update(*solutions.index(minute-1, uvalve_idx, valve_map[et], new_open_valves_idx) + (uvalve->flow_rate * minute));
                            }
                        }

                        // 4. you open, elephant opens
                        if(uvalve_idx != evalve_idx && uvalve->flow_rate > 0 && evalve->flow_rate > 0 && ((uvalve_bit | evalve_bit) & open_valves) == 0) {
                            uint16_t new_open_valves = open_valves | uvalve_bit | evalve_bit;
                            size_t new_open_valves_idx = possible_open_valves_idx_map[new_open_valves];
                            solution.maybe_update(*solutions.index(minute-1, uvalve_idx, evalve_idx, new_open_valves_idx) + (uvalve->flow_rate * minute) + (evalve->flow_rate * minute));
                        }
                    }
                }
            }
        }
    }

    uint16_t solution = *solutions.index(25, valve_map["AA"], valve_map["AA"], 0);

    cout << "Part 2: " << solution << endl;

    return 0;
}

File: day_16/solution.py
from pathlib import Path
import re
from itertools import chain, combinations

from tqdm import tqdm

USE_SAMPLE_INPUT = True
SKIP_PART_1 = True
in_path = Path.cwd() / 'input.txt'

sample_input = """
Valve AA has flow rate=0; tunnels lead to valves DD, II, BB
Valve BB has flow rate=13; tunnels lead to valves CC, AA
Valve CC has flow rate=2; tunnels lead to valves DD, BB
Valve DD has flow rate=20; tunnels lead to valves CC, AA, EE
Valve EE has flow rate=3; tunnels lead to valves FF, DD
Valve FF has flow rate=0; tunnels lead to valves EE, GG
Valve GG has flow rate=0; tunnels lead to valves FF, HH
Valve HH has flow rate=22; tunnel leads to valve GG
Valve II has flow rate=0; tunnels lead to valves AA, JJ
Valve JJ has flow rate=21; tunnel leads to valve II
""".strip()

if USE_SAMPLE_INPUT:
    input = sample_input
else:
    with in_path.open("r") as inf:
        input = inf.read().strip()


def powerset(iterable):
    "powerset([1,2,3]) --> () (1,) (2,) (3,) (1,2) (1,3) (2,3) (1,2,3)"
    s = list(iterable)
    return chain.from_iterable(combinations(s, r) for r in range(len(s)+1))

class Valve:
    STR_RE = re.compile(r'Valve (\S+) has flow rate=(\d+); tunnels? leads? to valves? (.*)$')

    def __init__(self, label: str, flow_rate: int, tunnels: 'list[str]'):
        self.label = label
        self.flow_rate = flow_rate
        self.tunnels = tunnels

    @staticmethod
    def from_str(inputstr: str):
        match = Valve.STR_RE.match(inputstr)
        if match is None:
            raise ValueError(f'Invalid valve string:"{inputstr}"')
        label = match.group(1)
        flow_rate = int(match.group(2))
        tunnels = match.group(3).split(", ")
        return Valve(label, flow_rate, tunnels)

    def __str__(self):
        return f'Valve {self.label} has flow rate={self.flow_rate}; tunnels lead to valves ' + ', '.join(self.tunnels)

    def __eq__(self, other):
        if not isinstance(other, Valve):
            return False
        return self.label == other.label and self.flow_rate == other.flow_rate and self.tunnels == other.tunnels

    def __hash__(self):
        return self.label.__hash__()

    def __repr__(self):
        return f"Valve({self.label})"

valves: 'dict[str, Valve]' = {}

for line in input.split('\n'):
    valve = Valve.from_str(line)
    valves[valve.label] = valve

for valve in valves.values():
    print(valve)

significant_valves = [valve for valve in valves.values() if valve.flow_rate > 0]
possible_open_valves = [frozenset(s) for s in powerset(significant_valves)]

if not SKIP_PART_1:

    solutions: 'list[dict[str, dict[set[Valve], int]]]' = []

    for i in trange(30):
        solutions.append({})
        #if i > 2:
        #    solutions[i-2] = []

        for valve in valves.values():
            solutions[i][valve.label] = {}
            for open_valves in possible_open_valves:
                sol = 0

                if i > 0:
                    for tunnel in valve.tunnels:
                        sol = max(sol, solutions[i-1][tunnel][open_valves])

                    if valve.flow_rate > 0 and valve not in open_valves:
                        new_open_valves = frozenset(open_valves.union([valve]))
                        sol = max(sol, solutions[i-1][valve.label][frozenset(new_open_valves)] + (valve.flow_rate * i))
                solutions[i][valve.label][open_valves] = sol

    solution = solutions[29]['AA'][frozenset()]
    print("Part 1:", solution)

    del solution
    del solutions
solutions: 'list[dict[str, dict[str, dict[set[Valve], int]]]]' = []
with tqdm(total=26 * len(valves) * len(valves) * len(possible_open_valves)) as pbar:
    for i in range(26):
        solutions.append({})
        if i > 2:
            solutions[i-2] = []

        for uvalve in valves.values():
            solutions[i][uvalve.label] = {}
            for evalve in valves.values():
                solutions[i][uvalve.label][evalve.label] = {}
                for open_valves in possible_open_valves:
                    pbar.update()
                    sol = 0

                    if i > 0:
                        # 1. you move, elephant moves
                        for ut in uvalve.tunnels:
                            for et in evalve.tunnels:
                                sol = max(sol, solutions[i-1][ut][et][open_valves])

                        # 2. you move, elephant opens
                        if evalve.flow_rate > 0 and evalve not in open_valves:
                            for ut in uvalve.tunnels:
                                new_open_valves = frozenset(open_valves.union([evalve]))
                                sol = max(sol, solutions[i-1][ut][evalve.label][new_open_valves] + (evalve.flow_rate * i))

                        # 3. you open, elephant moves
                        if uvalve.flow_rate > 0 and uvalve not in open_valves:
                            for et in evalve.tunnels:
                                new_open_valves = frozenset(open_valves.union([uvalve]))
                                sol = max(sol, solutions[i-1][uvalve.label][et][new_open_valves] + (uvalve.flow_rate * i))

                        # 4. you open, elephant opens
                        if uvalve.flow_rate > 0 and evalve.flow_rate > 0 and uvalve not in open_valves and evalve not in open_valves and uvalve != evalve:
                            new_open_valves = frozenset(open_valves.union([uvalve, evalve]))
                            sol = max(sol, solutions[i-1][uvalve.label][evalve.label][new_open_valves] + (uvalve.flow_rate * i) + (evalve.flow_rate * i))
                    solutions[i][uvalve.label][evalve.label][open_valves] = sol

def backtrace(moves_left: int, your_pos: str, elephant_pos: str, open_valves: 'set[Valve]'):
    print(f"== Minute {25-moves_left} ==")
    if len(open_valves) == 0:
        print("No valves are open.")
    else:
        print("Valves " + ", ".join(sorted([v.label for v in open_valves])) + " are open, releasing " + str(sum([v.flow_rate for v in open_valves])) + " pressure.")

    solution = solutions[moves_left][your_pos][elephant_pos][open_valves]

    if moves_left == 0:
        print()
        return

    print("Solution:", solution)

    uvalve = valves[your_pos]
    evalve = valves[elephant_pos]

    # 1. you move, elephant moves
    for ut in uvalve.tunnels:
        for et in evalve.tunnels:
            if solution == solutions[moves_left-1][ut][et][open_valves]:
                print(f"You move to valve {ut}.")
                print(f"The elephant moves to valve {et}.")
                print()
                backtrace(moves_left-1, ut, et, open_valves)
                return

    # 3. you open, elephant moves
    if uvalve.flow_rate > 0 and uvalve not in open_valves:
        for et in evalve.tunnels:
            new_open_valves = frozenset(open_valves.union([uvalve]))
            if solution == solutions[moves_left-1][uvalve.label][et][new_open_valves] + (uvalve.flow_rate * moves_left):
                print(f"You open valve {uvalve.label}.")
                print(f"The elephant moves to valve {et}.")
                print()
                backtrace(moves_left-1, uvalve.label, et, new_open_valves)
                return

    # 2. you move, elephant opens
    if evalve.flow_rate > 0 and evalve not in open_valves:
        for ut in uvalve.tunnels:
            new_open_valves = frozenset(open_valves.union([evalve]))
            if solution == solutions[moves_left-1][ut][evalve.label][new_open_valves] + (evalve.flow_rate * moves_left):
                print(f"You move to valve {ut}.")
                print(f"The elephant opens valve {evalve.label}.")
                print()
                backtrace(moves_left-1, ut, evalve.label, new_open_valves)
                return

    # 4. you open, elephant opens
    if uvalve.flow_rate > 0 and evalve.flow_rate > 0 and uvalve not in open_valves and evalve not in open_valves and uvalve != evalve:
        new_open_valves = frozenset(open_valves.union([uvalve, evalve]))
        if solution == solutions[moves_left-1][uvalve.label][evalve.label][new_open_valves] + (uvalve.flow_rate * moves_left) + (evalve.flow_rate * moves_left):
            print(f"You open valve {uvalve.label}.")
            print(f"The elephant opens valve {evalve.label}.")
            print()
            backtrace(moves_left-1, uvalve.label, evalve.label, new_open_valves)
            return

    raise ValueError("Should not be reached!!")

solution = solutions[-1]['AA']['AA'][frozenset()]

# backtrace(25, 'AA', 'AA', frozenset())

print("Part 2:", solution)
